src/asynk/async_scheduler.rs

```diff
@@ -0,0 +1,207 @@
+                    self.number_of_restarts
+                );
+                self.number_of_restarts += 1;
+                self.start().await
+            }
+        }
+    }
+
+    pub async fn schedule_loop(&mut self) -> Result<(), Error> {
+        let sleep_duration = Duration::from_secs(self.check_period);
+
+        loop {
+            self.schedule().await?;
+
+            sleep(sleep_duration).await;
+        }
+    }
+
+    pub async fn schedule(&mut self) -> Result<(), Error> {
+        if let Some(tasks) = self
+            .queue
+            .fetch_periodic_tasks(self.error_margin_seconds as i64)
+            .await?
+        {
+            for task in tasks {
+                self.process_task(task).await?;
+            }
+        };
+        Ok(())
+    }
+
+    async fn process_task(&mut self, task: PeriodicTask) -> Result<(), Error> {
+        match task.scheduled_at {
+            None => {
+                self.queue.schedule_next_task(task).await?;
+            }
+            Some(_) => {
+                let metadata = task.metadata.clone();
+                // known issue,  hardcode common to try if rest of code works
+                self.queue.insert_task(metadata, "common").await?;
+
+                self.queue.schedule_next_task(task).await?;
+            }
+        }
+        Ok(())
+    }
+
+    #[cfg(test)]
+    async fn schedule_test(&mut self) -> Result<(), Error> {
+        let sleep_duration = Duration::from_secs(self.check_period);
+
+        loop {
+            match self
+                .queue
+                .fetch_periodic_tasks(self.error_margin_seconds as i64)
+                .await?
+            {
+                Some(tasks) => {
+                    for task in tasks {
+                        self.process_task(task).await?;
+                    }
+
+                    return Ok(());
+                }
+                None => {
+                    sleep(sleep_duration).await;
+                }
+            };
+        }
+    }
+}
+
+#[cfg(test)]
+mod async_scheduler_tests {
+    use super::Scheduler;
+    use crate::asynk::async_queue::AsyncQueueTest;
+    use crate::asynk::async_queue::AsyncQueueable;
+    use crate::asynk::async_queue::PeriodicTask;
+    use crate::asynk::AsyncRunnable;
+    use crate::asynk::Error;
+    use async_trait::async_trait;
+    use bb8_postgres::bb8::Pool;
+    use bb8_postgres::tokio_postgres::NoTls;
+    use bb8_postgres::PostgresConnectionManager;
+    use chrono::DateTime;
+    use chrono::Duration as OtherDuration;
+    use chrono::Utc;
+    use serde::{Deserialize, Serialize};
+
+    #[derive(Serialize, Deserialize)]
+    struct AsyncScheduledTask {
+        pub number: u16,
+    }
+
+    #[typetag::serde]
+    #[async_trait(?Send)]
+    impl AsyncRunnable for AsyncScheduledTask {
+        async fn run(&self, _queueable: &mut dyn AsyncQueueable) -> Result<(), Error> {
+            Ok(())
+        }
+        /*fn task_type(&self) -> String {
```

Comment:
we will delete comment here when task type is fixed.

Reply:
Solve in [a689d02](https://github.com/ayrat555/fang/pull/47/commits/a689d022634d4fd8ba3e4690dc1557b24030bc00)